File: app/src/main/java/com/tentwenty/movieticket/feature/seatselection/SquareTextView.java
package com.tentwenty.movieticket.feature.seatselection;

import android.content.Context;
import android.graphics.Typeface;
import android.support.v4.content.ContextCompat;
import android.support.v7.widget.AppCompatTextView;
import android.util.AttributeSet;
import android.view.Gravity;
import android.view.MotionEvent;

import com.tentwenty.movieticket.R;


public class SquareTextView extends AppCompatTextView {

    private int mTextType = SeatSelectionActivity.SEAT_AVAILABLE;

    public SquareTextView(Context context) {
        super(context);
        init(null);
    }

    public SquareTextView(Context context, int textType) {
        super(context);
        mTextType = textType;
        init(null);
    }

    public SquareTextView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(attrs);
    }

    public SquareTextView(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        init(attrs);
    }

    private void init(AttributeSet attrs) {

        setBackgroundDrawable(ContextCompat.getDrawable(getContext(), R.drawable.text_bg));
        setAlphaOnTextColor();
        setText(getText());
        setTypeface(Typeface.DEFAULT_BOLD);
        setGravity(Gravity.CENTER);
    }

    void setAlphaOnTextColor() {
        switch (mTextType) {
            case SeatSelectionActivity.SEAT_AVAILABLE:
                setTextColor(ContextCompat.getColor(getContext(), android.R.color.darker_gray));
                setBackgroundDrawable(ContextCompat.getDrawable(getContext(), R.drawable.text_bg));
                break;
            case SeatSelectionActivity.SEAT_UNAVAILABLE:
                setTextColor(ContextCompat.getColor(getContext(), android.R.color.white));
                setBackgroundDrawable(ContextCompat.getDrawable(getContext(), R.drawable.text_bg_unavailable));

                break;
            case SeatSelectionActivity.EMPTY_SEAT_PASSAGE:
                setTextColor(ContextCompat.getColor(getContext(), android.R.color.transparent));
                setBackgroundColor(ContextCompat.getColor(getContext(),android.R.color.transparent));
                break;
            case SeatSelectionActivity.SEAT_SELECTED:
                setTextColor(ContextCompat.getColor(getContext(),  android.R.color.white));
                setBackgroundColor(ContextCompat.getColor(getContext(),android.R.color.holo_green_dark));
                break;
        }
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
        int width = this.getMeasuredWidth();
        int height = this.getMeasuredHeight();
        int size = Math.max(width, height);
        int widthSpec = MeasureSpec.makeMeasureSpec(size, MeasureSpec.EXACTLY);
        int heightSpec = MeasureSpec.makeMeasureSpec(size, MeasureSpec.EXACTLY);
        super.onMeasure(widthSpec, heightSpec);

    }

    public void setTextType(int mTextType) {
        this.mTextType = mTextType;
        requestLayout();
    }
}

File: app/src/main/java/com/tentwenty/movieticket/feature/seatselection/SquareTextView.kt
package com.tentwenty.movieticket.feature.seatselection

import android.content.Context
import android.graphics.Typeface
import android.support.v4.content.ContextCompat
import android.support.v7.widget.AppCompatTextView
import android.util.AttributeSet
import android.view.Gravity
import android.view.View

import com.tentwenty.movieticket.R


class SquareTextView : AppCompatTextView {

    private var mTextType = SeatSelectionActivity.SEAT_AVAILABLE

    constructor(context: Context) : super(context) {
        init(null)
    }

    constructor(context: Context, textType: Int) : super(context) {
        mTextType = textType
        init(null)
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        init(attrs)
    }

    constructor(context: Context, attrs: AttributeSet, defStyle: Int) : super(context, attrs, defStyle) {
        init(attrs)
    }

    private fun init(attrs: AttributeSet?) {

        setBackgroundDrawable(ContextCompat.getDrawable(context, R.drawable.text_bg))
        setAlphaOnTextColor()
        text = text
        typeface = Typeface.DEFAULT_BOLD
        gravity = Gravity.CENTER
    }

    internal fun setAlphaOnTextColor() {
        when (mTextType) {
            SeatSelectionActivity.SEAT_AVAILABLE -> {
                setTextColor(ContextCompat.getColor(context, android.R.color.darker_gray))
                setBackgroundDrawable(ContextCompat.getDrawable(context, R.drawable.text_bg))
            }
            SeatSelectionActivity.SEAT_UNAVAILABLE -> {
                setTextColor(ContextCompat.getColor(context, android.R.color.white))
                setBackgroundDrawable(ContextCompat.getDrawable(context, R.drawable.text_bg_unavailable))
            }
            SeatSelectionActivity.EMPTY_SEAT_PASSAGE -> {
                setTextColor(ContextCompat.getColor(context, android.R.color.transparent))
                setBackgroundColor(ContextCompat.getColor(context, android.R.color.transparent))
            }
            SeatSelectionActivity.SEAT_SELECTED -> {
                setTextColor(ContextCompat.getColor(context, android.R.color.white))
                setBackgroundColor(ContextCompat.getColor(context, android.R.color.holo_green_dark))
            }
            SeatSelectionActivity.SEAT_BOOKED -> {
                setTextColor(ContextCompat.getColor(context, android.R.color.white))
                setBackgroundDrawable(ContextCompat.getDrawable(context, R.drawable.text_bg_booked))
            }
        }
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        val width = this.measuredWidth
        val height = this.measuredHeight
        val size = Math.max(width, height)
        val widthSpec = View.MeasureSpec.makeMeasureSpec(size, View.MeasureSpec.EXACTLY)
        val heightSpec = View.MeasureSpec.makeMeasureSpec(size, View.MeasureSpec.EXACTLY)
        super.onMeasure(widthSpec, heightSpec)

    }

    fun setTextType(mTextType: Int) {
        this.mTextType = mTextType
        requestLayout()
    }
}